src/docker/functions/image/image-exists.ts

```diff
@@ -1,9 +1,25 @@
 import { DockerImageName } from "../../../docker-image-name";
-import { listImages } from "./list-images";
 import { log } from "../../../logger";
 import Dockerode from "dockerode";
+import AsyncLock from "async-lock";
+import { listImages } from "./list-images";
+
+const existingImages = new Set<string>();
+const imageCheckLock = new AsyncLock();
 
 export const imageExists = async (dockerode: Dockerode, imageName: DockerImageName): Promise<boolean> => {
   log.debug(`Checking if image exists: ${imageName}`);
-  return (await listImages(dockerode)).some((image) => image.equals(imageName));
+
+  return imageCheckLock.acquire(imageName.toString(), async () => {
```

Comment:
Is it necessary to provide a timeout here, in case something goes wrong and the lock can't be acquired? Likewise does the lock need to be released in some `finally` block?

Reply:
The lock implementation handles the [release in case the user code throws](https://github.com/rogierschouten/async-lock/blob/master/lib/index.js#L150-L179), it also allows configuring a timeout, but for now I left it out, as presumably the underlying operations already have timeouts?